features/190-consistent-error/api.md

```diff
@@ -0,0 +1,95 @@
+# Consistent error handling
+
+## Requirement
+
+* To support a consistent way of handling error
+* To obtain error information from failed API request
+* To improve TypeScript support with regards to data obtained from rejected
+  promise.
+* Support legacy browsers (aka pre-IE-11).
+
+## Current problem
+
+The main problem lies in how the JS SDK handle error. Depending on the type of
+API called, a promise maybe rejected with one of the followings:
+
+* Error object
+* SkygearError object (subclass of Error)
+* Plain object
+
+In the third case, the error message may be available in one of these locations:
+
+* `error.error.message` 
+* `error.message`
+
+## Proposed solution
+
+* Rejected promise should return an Error or an Error subclass. When calling
+  third party library, make sure the rejected promise returns an Error.
+* Skygear SDK must wrap server-side error object in SkygearError, which is
+  a subclass of Error.
+* For other kinds of error, it is preferred to wrap such errors into
+  a SkygearError and use an appropriate error code. Reject the promise
+  with a SkygearError.
+
+### Saving multiple records
+
+When saving multiple records, the returned error depends on how many records are
+being saved and whether the operation is atomic.
+
+To ensure a consistent return value and error for different cases, multiple save
```

Comment:
I have an impression that Javascript folks are ok with inconsistent return value, so we should not really need `saveMultiple`? Or is it because of TypeScript or I just got it wrong?

Reply:
Well, I guess if we have inconsistent return value, users might be caught off guard when the return value is not what they expected.

TypeScript is also a big reason we should have consistent return value.